libmisc/chowndir.c

```diff
@@ -65,68 +48,34 @@ int chown_tree (const char *root,
 	 * recursively.  If not, it is checked to see if an ownership
 	 * shall be changed.
 	 */
-
-	dir = opendir (root);
-	if (NULL == dir) {
-		free (new_name);
-		return -1;
-	}
-
 	while ((ent = readdir (dir))) {
-		size_t ent_name_len;
 		uid_t tmpuid = (uid_t) -1;
 		gid_t tmpgid = (gid_t) -1;
+		struct stat ent_sb;
 
 		/*
 		 * Skip the "." and ".." entries
 		 */
-
 		if (   (strcmp (ent->d_name, ".") == 0)
 		    || (strcmp (ent->d_name, "..") == 0)) {
 			continue;
 		}
 
-		/*
-		 * Make the filename for both the source and the
-		 * destination files.
-		 */
-
-		ent_name_len = strlen (root) + strlen (ent->d_name) + 2;
-		if (ent_name_len > new_name_len) {
-			/*@only@*/char *tmp = realloc (new_name, ent_name_len);
-			if (NULL == tmp) {
-				rc = -1;
-				break;
-			}
-			new_name = tmp;
-			new_name_len = ent_name_len;
-		}
-
-		(void) snprintf (new_name, new_name_len, "%s/%s", root, ent->d_name);
-
-		/* Don't follow symbolic links! */
-		if (LSTAT (new_name, &sb) == -1) {
-			continue;
+		rc = fstatat (dirfd(dir), ent->d_name, &ent_sb, AT_SYMLINK_NOFOLLOW);
+		if (rc < 0) {
+			break;
 		}
 
-		if (S_ISDIR (sb.st_mode) && !S_ISLNK (sb.st_mode)) {
-
+		if (S_ISDIR (ent_sb.st_mode)) {
 			/*
 			 * Do the entire subdirectory.
 			 */
-
-			rc = chown_tree (new_name, old_uid, new_uid,
-			                 old_gid, new_gid);
+			rc = chown_tree_at (dirfd(dir), ent->d_name, old_uid, new_uid, old_gid, new_gid);
```

Comment:
So as long as you're doing this, should you also be opening the file at the top of loop and using fstat and fchown?

Reply:
Pinning the directory is needed to avoid it being replaced with a symlink, which we then would follow.
Pinning the directory entries would avoid them being replaced by a different kind of file (between fstatat() and fchownat()), but since we do not follow symlinks for these operations there is no modification on arbitrary files.
(Also since fchown() does not support O_PATH file descriptors, we would need to open the files with O_RDONLY, which e.g. in case of SELinux requires elevated permissions.)

Reply:
I could still imagine a race leaking data owned by user1 to user2, but user2 couldn't cause that, only benefit from it, so this is probably ok.  Thanks.